core/primitives/src/utils.rs

```diff
@@ -469,23 +472,87 @@ where
     Serializable(object)
 }
 
-/// Derives `AccountId` from `PublicKey``.
+/// From `near-account-id` version `1.0.0-alpha.2`, `is_implicit` returns true for ETH-implicit accounts.
+/// This function is a wrapper for `is_implicit` method so that we can easily differentiate its behavior
+/// based on the protocol version.
+pub fn account_is_implicit(account_id: &AccountId, protocol_version: ProtocolVersion) -> bool {
+    if checked_feature!("stable", EthImplicit, protocol_version) {
+        account_id.get_account_type().is_implicit()
+    } else {
+        account_id.get_account_type() == AccountType::NearImplicitAccount
+    }
+}
+
+/// Derives `AccountId` from `PublicKey`.
 /// If the key type is ED25519, returns hex-encoded copy of the key.
-pub fn derive_near_implicit_account_id(public_key: &ED25519PublicKey) -> AccountId {
-    hex::encode(public_key).parse().unwrap()
+/// If the key type is SECP256K1, returns '0x' + keccak256(public_key)[12:32].hex().
+pub fn derive_account_id_from_public_key(public_key: &PublicKey) -> AccountId {
+    match public_key.key_type() {
+        KeyType::ED25519 => hex::encode(public_key.key_data()).parse().unwrap(),
+        KeyType::SECP256K1 => {
+            use sha3::Digest;
+            let pk_hash = sha3::Keccak256::digest(&public_key.key_data());
+            format!("0x{}", hex::encode(&pk_hash[12..32])).parse().unwrap()
+        }
+    }
+}
+
+// TODO(eth-implicit) This functionality will be managed by `Wallet Contract`
+// and no access key will be added.
```

Comment:
Without the `Wallet Contract` placeholder I would have to remove perhaps the entire code from this PR to keep CI happy. I think there are 2 options:
1. We merge this PR as a basis for the next PR. 
2. I close this PR and work on a new one with `Wallet Contract` in mind from the beginning.

I have no strong opinion which approach is better.
The first option requires more reviewers time but it can spot some issues earlier (by reviewers or nightly testing).
The second option looks clearer, as it be better visible within one PR what changes, and save reviewers time now.
Feel free to just leave some comments, that would help in both options.

Reply:
I like the second approach better. Not just to save review time, but to save you time as well. I think starting fresh from the master branch should be easier for the new PR than trying to remove code you just added.

The overall structure of this PR looks good to me. I don't have any comments about this PR which might inform you for the new PR.